c/core.go

```
@@ -0,0 +1,173 @@
+package c
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+)
+
+// WithRestart specifies how the parent supervisor should restart this child
+// after an error is encountered.
+func WithRestart(r Restart) Opt {
+	return func(spec *Spec) {
+		spec.restart = r
+	}
+}
+
+// WithShutdown specifies how the shutdown if the child is going to be handled.
+func WithShutdown(s Shutdown) Opt {
+	return func(spec *Spec) {
+		spec.shutdown = s
+	}
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+// New creates an Spec for a worker goroutine.
+//
+// It requires two arguments:
+//
+// - name: The name of the worker (for tracing purposes)
+//
+// - start: The main function that is going to be executing in this child
+// goroutine
+//
+// The start function will receive a context.Context record that must be used to
+// receive a stop signal from the parent supervisor. Depending on the Shutdown
+// settings, if the start function does not respect the given context, the
+// supervision system will either block forever or leak goroutines after a
+// timeout has been reached.
+func New(name string, start func(context.Context) error, opts ...Opt) Spec {
+	return New1(name, func(ctx context.Context, notifyChildStart func()) error {
+		notifyChildStart()
+		return start(ctx)
+	}, opts...)
+}
+
+// New1 is similar to New, with the difference that the start function receives
+// a `notifyStart` callback to the child routine, this callback allows the child
+// goroutine to signal when it has officially started. This is essential when
+// you want to guarantee some bootstrap on thread initialization.
+func New1(
+	name string,
+	start func(context.Context, func()) error,
+	opts ...Opt,
+) Spec {
+	spec := Spec{}
+
+	if name == "" {
+		panic("Child cannot have empty name")
```

Add comment explaining rationale of panic instead of returning error